src/zap/zapimport.cpp

```diff
@@ -1993,238 +2003,268 @@ DWORD ZapIndirectHelperThunk::SaveWorker(ZapWriter * pZapWriter)
-        *(DWORD *)p = 0xC000F8DC;
-        p += 4;
+            // move Module* -> x10
+            // ldr x10, [PC+0x14]
+            *(DWORD*)p = 0x580000AA;
+            p += 4;
 
-        // bx r12
-        *(WORD *)p = 0x4760;
-        p += 2;
-    }
-#elif defined(_TARGET_ARM64_)
-    if (IsDelayLoadHelper())
-    {
-        // x11 contains indirection cell
-        // Do nothing x11 contains our first param
+            //ldr x10, [x10]
+            *(DWORD*)p = 0xf940014A;
+            p += 4;
+        }
+        else
+        if (IsLazyHelper())
+        {
+            // Move Module* -> x1
+            // ldr x1, [PC+0x14]
+            *(DWORD*)p = 0x580000A1;
+            p += 4;
 
-        //  movz x9, #index
-        DWORD index = GetSectionIndex();
-        _ASSERTE(index <= 0x7F);
-        *(DWORD*)p = 0xd2800009 | (index << 5);
-        p += 4;
+            // ldr x1, [x1]
+            *(DWORD*)p = 0xf9400021;
+            p += 4;
+        }
 
-        // move Module* -> x10
-        // ldr x10, [PC+0x14]
-        *(DWORD*)p = 0x580000AA;
+        // branch to helper
+        // ldr x12, [PC+0x14]
+        *(DWORD*)p = 0x580000AC;
         p += 4;
 
-        //ldr x10, [x10]
-        *(DWORD*)p = 0xf940014A;
-        p += 4;
-    }
-    else
-    if (IsLazyHelper())
-    {
-        // Move Module* -> x1
-        // ldr x1, [PC+0x14]
-        *(DWORD*)p = 0x580000A1;
+        // ldr x12, [x12]
+        *(DWORD *)p = 0xf940018c;
         p += 4;
 
-        // ldr x1, [x1]
-        *(DWORD*)p = 0xf9400021;
+        // br x12
+        *(DWORD *)p = 0xd61f0180;
         p += 4;
-    }
 
-    // branch to helper
-    // ldr x12, [PC+0x14]
-    *(DWORD*)p = 0x580000AC;
-    p += 4;
+        // [Module*]
+        if (pImage != NULL)
+            pImage->WriteReloc(buffer, (int)(p - buffer), pImage->GetImportTable()->GetHelperImport(READYTORUN_HELPER_Module), 0, IMAGE_REL_BASED_PTR);
+        p += 8;
 
-    // ldr x12, [x12]
-    *(DWORD *)p = 0xf940018c;
-    p += 4;
+        // [helper]
+        if (pImage != NULL)
+            pImage->WriteReloc(buffer, (int)(p - buffer), pImage->GetImportTable()->GetHelperImport(GetReadyToRunHelper()), 0, IMAGE_REL_BASED_PTR);
+        p += 8;
 
-    // br x12
-    *(DWORD *)p = 0xd61f0180;
-    p += 4;
+#else
+        PORTABILITY_ASSERT("ZapIndirectHelperThunk::SaveWorker");
+#endif
 
-    // [Module*]
-    if (pImage != NULL)
-        pImage->WriteReloc(buffer, (int)(p - buffer), pImage->GetImportTable()->GetHelperImport(READYTORUN_HELPER_Module), 0, IMAGE_REL_BASED_PTR);
-    p += 8;
+        _ASSERTE(p - buffer <= sizeof(buffer));
+
+        if (pZapWriter != NULL)
+            pZapWriter->Write(&buffer, (int)(p - buffer));
+
+        return (DWORD)(p - buffer);
+    }
 
-    // [helper]
+#endif // FEATURE_READYTOTRUN_COMPILER
+
+    // for ngen images, we use this to write jit helper jumps through an indirection cell
+    BYTE buffer[6];
+    BYTE * p = buffer;
+
+#if defined(_TARGET_X86_)
```

Comment:
Can you please also implement this for X86 - ARM/ARM64 can wait, but it would nice to have this working on X86 and AMD64 (not because of this code, but because other things are rather different - for example compact entry points are X86 only).

Reply:
Will do as soon as I get a chance, probably next week.